aten/src/ATen/native/TensorShape.cpp

```diff
@@ -912,7 +912,7 @@ Tensor narrow_copy_sparse(const Tensor& self, int64_t dim, int64_t start, int64_
 }
 
 Tensor& narrow_copy_dense_cpu_out(
-  const Tensor& self, int64_t dim, int64_t start, int64_t length, Tensor& output
+  const Tensor& self, int64_t dim, int64_t start, SymbolicOrConcreteInt length, Tensor& output
```

Comment:
As I mentioned last week, kernels shouldn't take SymbolicOrConcreteInt as they cannot usefully operate with a symbolic int. So we'll wanna change these back to int64_t and teach codegen about the conversion

Reply:
(this is worth doing sooner rather than later because it will reduce the amount of changes like this you'll have to make)

Reply:
To clarify -- are you suggesting that:
1. all of {narrow_copy_dense, narrow_copy_dense_cpu, narrow_copy_dense_cpu_out, narrow_copy_sparse} shouldn't take SymbolicOrConcreteInt?
1. Or just "kernels" as in functions where the compute happens? 
1. Or do you mean functions that are not composite with respect to autograd?

I can see the case for (2), I'm not sure about (3)

Reply:
@ezyang , @zou3519 I have a few more qq as well

1. What do we want to happen if a user passes a true `SymbolicInt` to a non-lazy aten op?
2. Intuitively, it would seem to have the opposite effect. Instead of doing conversions + assert to int64_t in a callee (a concrete computation kernel), we will be pushing conversions to the callsites which can be inside of composite ops, user/model code, and in python. It sounds like a more invasive and tricky change.

Reply:
Kernels where compute happens. Like, any context where you need a concrete int, a symbolic int simply will not work (and will need to raise an error). The exact seam with regards to the codegen might require a little finagling but this is the high level.

> What do we want to happen if a user passes a true SymbolicInt to a non-lazy aten op?

Errors.

> Instead of doing conversions + assert to int64_t in a callee (a concrete computation kernel), we will be pushing conversions to the callsites which can be inside of composite ops, user/model code, and in python. It sounds like a more invasive and tricky change.

It's not. The signature of these kernels is code generated and the call sites are also code generated (e.g., you don't call at::native::narrow_copy_dense_cpu directly, you call at::narrow_copy). Thus the conversions can be automatically added (via codegen). I suppose you may not be so keen if you don't know how the codegen works, but this is a good opportunity to learn!

Reply:
For 1) ideally SymbolicInt only gets used when the input is a LazyTensor (or `__torch_dispatch__` object, in the case of AOTAutograd). Otherwise things are going to get a little weird and we will need to support "dispatching to the underlying backend of the SymbolicInt".

Do people mix and match LazyTensors and regular Tensors in torch/xla?

Reply:
> Do people mix and match LazyTensors and regular Tensors in torch/xla?

Generally no- we require all tensors to be lazy if any are. 

Exception: wrapping scalar-tensors automatically in lazy tensors, in some cases the nativefunc sig has a tensor not a scalar, but it has size {}, and we allow e.g. a cpu tensor for that arg to mingle with other lazy tensors (we wrap it).

Reply:
@ezyang 

Alright, it sounds like the main benefit would be to "hide" checked conversions to in64_t from the kernel code? 

This indeed would reduce the number of changes we have to make, but this change could probably be postponed a bit since 
a) we are only going to change a handful of kernels initially, so it won't be too hard to remove explicit conversions
b) also there are scarier pieces I'm anxious to figure out first such as adding a list of `SymbolicOrConcreteInt`s and making it work and how `SymbolicOrConcreteInt`s would interact with functionalization kernels.

Reply:
Yes, of course. It's good to have a design so we do not have to worry about losing things like this even though we have to implement other things first ;)

Reply:
Edward will post the code change to make the codegen do it